src/Shared/FileMatcher.cs

```diff
@@ -1094,215 +1119,281 @@ private static string RegularExpressionFromFileSpec
+                regex.Append(FileSpecRegexParts.DirSeparator);
+            }
+            else if (IsSpecialRegexCharacter(ch))
+            {
+                regex.Append('\\');
+                regex.Append(ch);
+            }
+            else
+            {
+                regex.Append(ch);
+            }
+        }
 
-            /*
-             * Remaining call-outs not involving "**"
-             */
-            matchFileExpression.Replace("*.*", "<:anynonseparator:>");
-            matchFileExpression.Replace("*", "<:anynonseparator:>");
-            matchFileExpression.Replace("?", "<:singlecharacter:>");
+        private static bool IsSpecialRegexCharacter(char ch) =>
+            ch == '\\' || ch == '$' || ch == '(' || ch == ')' || ch == '+'
+            || ch == '.' || ch == '[' || ch == '^' || ch == '{' || ch == '|';
 
-            /*
-             *  Escape all special characters defined for regular expresssions.
-             */
-            matchFileExpression.Replace("\\", "\\\\"); // Must be first.
-            matchFileExpression.Replace("$", "\\$");
-            matchFileExpression.Replace("(", "\\(");
-            matchFileExpression.Replace(")", "\\)");
-            matchFileExpression.Replace("*", "\\*");
-            matchFileExpression.Replace("+", "\\+");
-            matchFileExpression.Replace(".", "\\.");
-            matchFileExpression.Replace("[", "\\[");
-            matchFileExpression.Replace("?", "\\?");
-            matchFileExpression.Replace("^", "\\^");
-            matchFileExpression.Replace("{", "\\{");
-            matchFileExpression.Replace("|", "\\|");
+        /// <summary>
+        /// Given an index directly after a directory separator,
+        /// iteratively skip three cases involving directory separators:
+        ///
+        ///  (1) \.\ -> \
+        ///     This is an identity, so for example, these two are equivalent,
+        ///
+        ///         dir1\.\dir2 == dir1\dir2
+        /// 
+        ///     (2) \\ -> \
+        ///         Double directory separators are treated as a single directory separator,
+        ///         so, for example, this is an identity:
+        ///
+        ///             f:\dir1\\dir2 == f:\dir1\dir2
+        ///
+        ///         The single exemption is for UNC path names, like this:
+        ///
+        ///             \\server\share != \server\share
+        /// 
+        ///         This case is handled by isUncPath in
+        ///         a prior step.
+        ///
+        /// (3) \**\**\ -> \**\
+        ///              This is an identity, so for example, these two are equivalent,
+        ///
+        ///                 dir1\**\**\ == dir1\**\
+        /// </summary>
+        /// <returns>The first index of a non-collapsible character.</returns>
+        private static int IndexOfNextNonCollapsibleChar(string str, int startIndex)
+        {
+            if (startIndex > 0 && !FileUtilities.IsAnySlash(str[startIndex - 1]))
```

Comment:
Can `startIndex-1` never be `.` or `*`? I'd add a precondition to document this.

Reply:
Yup, what if the starting string is `./` or `**`, then the method wouldn't move the cursor over them because startindex is not a slash.